### PR TITLE
Fixes #1228 : Add ImageClickInput Module

### LINKED ISSUE #1228: Add ImageRegionSelect Module
Introduce the following things in order to add Image Region Selection Interaction support

- [ ] Annotation class for Image Region Selection Rules classfiers

- [ ] IsInRegion Classifier

- [ ] MultiTypeDoubleInputMatcher for matching Image Region Selection classifiers

- [ ] Declaring ImageRegionSelectionModule in application component

- [ ] Registering Interaction inside InteractionObjectTypeExtractorRepository

**Target PR date**: 12 June 2020
**Target completion date**: 16 June 2020

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
## Explanation

Fixes part of #1228 

- Add ClickOnImage Proto Model

- Add IsInRegion Classifier

- Add ImageClickInputModule 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/main/java/org/oppia/app/application/ApplicationComponent.kt [hunk 1]
```diff
@@ -36,7 +37,7 @@ import org.oppia.util.threading.DispatcherModule
     ContinueModule::class, FractionInputModule::class,
     ItemSelectionInputModule::class, MultipleChoiceInputModule::class,
     NumberWithUnitsRuleModule::class, NumericInputRuleModule::class,
-    TextInputRuleModule::class, DragDropSortInputModule::class,
+    TextInputRuleModule::class, DragDropSortInputModule::class, ImageClickInputModule::class,
```
Comment (vinitamurthi) [inline app/src/main/java/org/oppia/app/application/ApplicationComponent.kt]: Can you move this to the next line to maintain consistency in the way the modules are listed?

File: domain/src/main/java/org/oppia/domain/classify/InteractionsModule.kt [hunk 2]
```diff
@@ -87,4 +87,13 @@ class InteractionsModule {
   ): InteractionClassifier {
     return GenericInteractionClassifier(ruleClassifiers)
   }
+
+  @Provides
+  @IntoMap
+  @StringKey("ImageClickInput")
+  fun provideImageClickInputInteractionClassifier(
+    @DragDropSortInputRules ruleClassifiers: Map<String, @JvmSuppressWildcards RuleClassifier>
```
Comment (vinitamurthi) [inline domain/src/main/java/org/oppia/domain/classify/InteractionsModule.kt]: `@ImageClickInputRules` maybe?

File: domain/src/main/java/org/oppia/domain/classify/rules/imageClickInput/ImageClickInputIsInRegionRuleClassifierProvider.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  override fun createRuleClassifier(): RuleClassifier {
+    return classifierFactory.createMultiTypeSingleInputClassifier(
+      InteractionObject.ObjectTypeCase.CLICK_ON_IMAGE,
+      InteractionObject.ObjectTypeCase.NORMALIZED_STRING,
+      "x",
+      this
+    )
+  }
+
+  override fun matches(answer: ClickOnImage, input: String): Boolean {
```
Comment (vinitamurthi) [inline domain/src/main/java/org/oppia/domain/classify/rules/imageClickInput/ImageClickInputIsInRegionRuleClassifierProvider.kt:29]: Based on https://github.com/oppia/oppia/blob/37285a9b0e7866cbcdf5f8193a6ac2fab64458a7/extensions/interactions/ImageClickInput/directives/image-click-input-rules.service.ts#L29 it looks like inputs is of 'any' type, although I see a comment saying that they actually want to make it a specific type. Have we confirmed that the actual type is string only and not a complex object?

**Review by vinitamurthi — CHANGES_REQUESTED** [review]

File: model/src/main/proto/interaction_object.proto [hunk 4]
```diff
@@ -57,6 +58,11 @@ message ListOfSetsOfHtmlStrings {
   repeated StringList set_of_html_strings = 1;
 }
 
+message Point2d {
```
Comment (rt4914) [inline model/src/main/proto/interaction_object.proto:62]: Please add comments for this explaining what it is?

**Review by rt4914 — APPROVED** [review]
LGTM, just on nit change.

File: domain/src/main/java/org/oppia/domain/classify/InteractionsModule.kt (hunk 2 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/InteractionsModule.kt]: yes I've missed this

File: app/src/main/java/org/oppia/app/application/ApplicationComponent.kt (hunk 1 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/application/ApplicationComponent.kt]: will do

File: domain/src/main/java/org/oppia/domain/classify/rules/imageClickInput/ImageClickInputIsInRegionRuleClassifierProvider.kt (hunk 3 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/imageClickInput/ImageClickInputIsInRegionRuleClassifierProvider.kt:29]: yes I have confirmed this that this object is of type string only

File: app/src/main/java/org/oppia/app/application/ApplicationComponent.kt (hunk 1 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/application/ApplicationComponent.kt]: done

File: domain/src/main/java/org/oppia/domain/classify/InteractionsModule.kt (hunk 2 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/InteractionsModule.kt]: done

File: model/src/main/proto/interaction_object.proto (hunk 4 above)
Comment (aggarwalpulkit596) [reply model/src/main/proto/interaction_object.proto:62]: done

**Review by vinitamurthi — APPROVED** [review]

**Pull request merged**